### DISCUSSION TITLE
IL-specific C# implementation spec

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#8733] [Q&A]
Best I can tell, the C# language spec doesn't specify how custom operators are translated to and from IL.

I wanted an exhaustive list of all specially-recognized operator method names (e.g. `op_Modulus`). The only place any `op_*` names show up in the spec is in explaining how to target operators in [doc comments](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md).

 1. Would you please maintain a single table mapping C# operator definition syntax to IL method name?
 2. Since the language seems to be at least coupled to the IL implementation names via the doc comments spec, though the syntax/IL mapping is never _specified_, wouldn't such a table belong as a specification nearer the operators section and not as an afterthought for doc comments?

**Comment by HaloFour** [comment]
IL has no concept of custom operators.  The `op_Whatever` convention is simply one established and understood between the various compilers.  Even then not all compilers agree exactly on those operator conventions which is why some languages have operators that others don't.

**Comment by HaloFour** [comment]
Oh, nevermind.  Reading that again what you're looking for is that convention, at least as it relates to C#.

You might want to ask on the VB.NET and F# forums too as their answers will be slightly different.

**Comment by svick** [comment]
The ECMA CLI spec does contain informative mapping between operator names and C++ operators, in [§I.10.3 Operator overloading](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf#%5B%7B%22num%22%3A1168%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C87%2C435%2C0%5D).

Those C++ operators are largely the same as C# operators, but I understand if that's not good enough for you.

Maybe this shouldn't be part of the C# spec, but it could be part of [Roslyn's documentation](https://github.com/dotnet/roslyn/tree/master/docs/compilers)?

**Comment by AnthonyDGreen** [comment]
This might help: http://source.roslyn.io/#Microsoft.CodeAnalysis/Symbols/WellKnownMemberNames.cs,6d17843a3c182cb7

**Comment by tannergooding** [comment]
> IL has no concept of custom operators.

IL doesn't, but the runtime spec (ECMA 335) does. It actually defines a number of "well-defined names". Most of which are supported by C#, but several of which (such as `op_PointerDereference`) are not. You can see Section `I.10.3 Operator Overloading` for more details.

**Comment by BillWagner** [comment]
Reviewed, and not moving to dotnet/csharpstandard. This shouldn't be part of the language standard, but should be documented elsewhere.
